### DISCUSSION TITLE
Two-Robot Scenarios

### DISCUSSION THREAD

**Discussion by Kai-X-Org** [HumanSupportRobot/QandA#68] [Q&A]
Is it possible to simulate scenarios with two HSR robots? If so, how would we go about spawning and establishing control connections to the second robot? Thanks a lot!

**Comment by masudam** [comment]
It is possible to spawn two robots, but since detailed operation verification has not been performed, there is a possibility of issues arising when actually running it. 
Below are the steps:

1. Launch with the "personal_name" specified as an argument:
   ```
   roslaunch hsrb_gazebo_launch (world).launch personal_name:=(robot name)
   ```
   Example:
   ```
   roslaunch hsrb_gazebo_launch hsrb_megaweb2015_world.launch personal_name:=robot1
   ```

2. For the second robot and beyond, use the launch file that only spawns the robot:
   ```
   roslaunch hsrb_gazebo_launch spawn_hsr.launch personal_name:=(robot name) target_map:=(map) x:=(x coordinate) y:=(y coordinate) z:=(z coordinate) Y:=(yaw)
   ```
   Example:
   ```
   roslaunch hsrb_gazebo_launch spawn_hsr.launch personal_name:=robot2 target_map:=megaweb x:=-1.0 y:=-6.0 z:=0.5 Y:=0.0
   ```

Please refer to the provided worlds in the `hsrb_gazebo_launch` package and the maps in the `tmc_potential_maps` package. Adjust the parameters in RViz under Displays to match the "personal_name."

* Set the RobotModel's Robot Description to `(robot name)/robot_description`.
* Set the RobotModel's TF Prefix to `(robot name)`.
* Set each Topic to match `(robot name)`.

Please note that proper testing and verification are essential when using multiple robots, as untested scenarios may lead to unexpected behavior.

**Comment by Kai-X-Org** [comment]
Thanks a lot for the response!

How would we use the Python API under the multi-robot scenario? It seems like right now, for code like robot.try_get, we can only gain control of the robot that is present on start up. How will we be able to gain control on the parts of different robots? Thanks a lot!

**Reply by masudam** [reply]
The Python API does not have a built-in feature to change the topic names it uses, so it seems that you can't use multiple robots as-is.

By referring to the code inside, you might find that it's possible to make adjustments by changing the topic names if necessary.
https://github.com/orgs/HumanSupportRobot/discussions/65#discussioncomment-6655953

**Reply by Kai-X-Org** [reply]
Thanks a lot! I am currently on working to create two-robot scenes. However, upon spawning the second robot named robot2, publishing to the /robot2/goal topic yields no response. It seems that the return value of rospy.get_num_connections is 0. Other topics seems to work. It would be great if I may have some help on that. Thank you so much!

**Reply by yuka-hashiguchi** [reply]
Hi @Kai-X-Org, could you kindly share the steps to replicate? When I executed the following commands on my end, I confirmed that robot2 moved 0.5 meter forward. I hope I could be of some assistance.
```
# launch the simulator with robot1
# In this case, I have opted for a relatively compact environment
roslaunch hsrb_gazebo_launch hsrb_hcr2013_world.launch personal_name:=robot1

# spawn robot2
roslaunch hsrb_gazebo_launch spawn_hsr.launch personal_name:=robot2 target_map:=hcr2013 x:=0.0 y:=1.0 z:=0.5 Y:=0.0

# publish move_base goal for robot2
$ rostopic pub /robot2/goal geometry_msgs/PoseStamped "header:
  seq: 0
  stamp:
    secs: 0
    nsecs: 0
  frame_id: 'map'
pose:
  position:
    x: 0.5
    y: 1.0
    z: 0.0
  orientation:
    x: 0.0
    y: 0.0
    z: 0.0
    w: 1.0" 
```

**Comment by Kai-X-Org** [comment]
Thank you very much!
I did it with Python:

```
rospy.init_node('test')

  
pub = rospy.Publisher('/robot2/goal', PoseStamped, queue_size=10)


while pub.get_num_connections() < 2:
    rospy.sleep(0.1)

  
goal_x =  1.0
goal_y = 0.0
goal_yaw = 1.57

  
goal = PoseStamped()
goal.header.stamp = rospy.Time.now()
goal.header.frame_id = "map" # Let's try the world frame
goal.pose.position = Point(goal_x, goal_y, 0)
quat = tf.transformations.quaternion_from_euler(0, 0, goal_yaw)
goal.pose.orientation = Quaternion(*quat)

  
pub.publish(goal)
```

Hoever, the whole thing just gets stuck in the while loop. Some topics, such as the one that controls the arm and head, do work, but this navigation one does not.

**Reply by Kai-X-Org** [reply]
I also tried your code, and it indeed works. Is there something wrong on the Python end? Thanks a lot!

**Reply by yuka-hashiguchi** [reply]
Thank you for sharing your code. I think it will work if you set the condition of the while loop to `pub.get_num_connections() < 1` or `pub.get_num_connections() == 0`. By default, there is only one node(`/robot2/move_base`) subscribing to `/robot2/goal`, so `pub.get_num_connections()` will be 1.

**Reply by hawkina** [reply]
this might be a bit old, but the same solution should apply in your case also: https://answers.ros.org/question/41433/multiple-robots-simulation-and-navigation/

tl;dr you can use tf_prefix parameter in the launch file to prefix the two robot namespaces. You would need to do this for your two control instances also. This method is relevant if you have one ros_master and two robots using it, so it will most likely be applicable to your simulation scenario. If you would have two real HSR robots, you would either need to make them communicate via REST (or similar) or make one robot use the other one as ROS_MASTER_URI. Both approaches have their up and down sides. Just be careful to really make sure, that every rostopic is prefixed correctly. 
If the topic name is hardcoded in the code and not named in the launch file, the tf_prefix might not work (I am not sure exactly when this case applies, but it can happen, therefore doublecheck)
Good luck!
